### PR TITLE
Reimplement library on top of `rustls::kernel` API (continuation of #61)

### LINKED ISSUE #59: Support new rustls `KernelConnection` API, and thence TLS1.3 KeyUpdates
See https://github.com/rustls/rustls/issues/2362 for background. I plan to work on this as part of integrating KTLS into  rustls/rustls-openssl-compat.

### PR DESCRIPTION
**Pull request by cxw620**
This PR inherits from #61 (#61 has had no activities for nearly two months, but I could not wait for it :) )
 
This PR completely rewrites this crate using the `rustls::kernel` API provided by rustls since rustls 0.23.27.

## Breaking changes

Almost everything.

The MSRV is now 1.77.0.

## Questions

1. Is this crate a low-level crate, or a high-level crate (similar to `tokio-rustls`)? Should the example implementations of (Ktls)Connector / (Ktls)Acceptor be merged into this crate, or just left them to be implemented by `tokio-rustls` (or any other crate)?

   My opinion is that we could create a new crate called `ktls-util`, implementing KtlsConnector / KtlsAcceptor, provided as an optional feature. Let higher-level crate `tokio-rustls` implement the fallback mechanism to transparently provide best-effort kTLS offload support for applications.

2. For detecting `CompatibleCiphers`, should it be implemented within this crate?

   My opinion is that this could also be implemented in `ktls-util`, provided as an optional feature.

## TODOs

- Improves testing.

  Due to the difficulty of enumerating all edge cases in real-world scenarios and the complexity of writing tests, more investigation is needed.

  (Additionally, I will privately deploy this in actual environments for testing)

- Reviews whether the current implementation complies with RFC 8446 and other relevant RFC document requirements.
- ~~More comprehensive documentation.~~ Done
- ~~Replaces `slicur::Reader` with `rustls::internal::msgs::codec::Reader`~~ Done.
- ~~Misc, such as applying cargo clippy / rustfmt, etc.~~ Done

---

Looking forward to feedback and code review.

---

Fixed: #59

### CODE REVIEW

**Comment by cxw620** [comment]
> I have put this back to draft. Please work on making this reviewable, by (for example) minimising back-and-forth changes, making a sequence of logical commits, and perhaps slicing off some smaller PRs that are easier to review and land, but still make sense individually.

Hi, I've cleaned up and sorted the commits.

Since almost everything is rewritten, I think we can hardly break this PR into some smaller ones.

**Comment by djc** [comment]
> Since almost everything is rewritten, I think we can hardly break this PR into some smaller ones.

If that's not possible I don't think any of the rustls maintainers will be able to review it. Maybe @fasterthanlime will be able to?

**Comment by cxw620** [comment]
> The point in making smaller commits is seeing how you are evolving the current code towards the new desired state. That is usually a lot easier then closely reviewing "here's a bunch of new code I wrote" after deleting all the old code.
> 
> As such, the current state of the PR isn't any easier for me to review than the previous state.

I attempted to refactor this crate incrementally, as you requested, while keeping the crate compilable at each step, preserving some of the old code to show "how I am evolving the current code towards the new desired state." However, I was unable to achieve this goal. Writing glue code for deprecated and soon-to-be-removed code is really frustrating.

I don't mean to criticize anyone or complain. I understand that this PR introduces sweeping changes that make it difficult to review. Perhaps I should consider creating a new crate on my own instead, if the commits I recently force-pushed still cannot meet your requirements.

**Comment by swlynch99** [comment]
Sorry for disappearing on the last PR and thanks @cxw620 for picking this up! I'm glad this didn't get stalled completely due to me getting burned out over the summer and I like the new work you've done to finish it :)

Here's some suggestions for things you can break into separate PRs that should help the maintainers review things more easily:
- Create a separate PR with mostly just the crate reorganization changes (e.g. creating the `setup`) module. Possibly multiple if possible.
- The rewrite of the code to probe for supported protocols can be moved to its own PR. The original `KtlsStream` struct doesn't interact with this in any way so it should be possible to do this without affecting much else.
- I'm not sure that there's a good way to break `KtlsStream` apart so this might be one big PR, but it may be possible to put it in as a skeleton and build it up using multiple PRs.

I'm going to take a crack at splitting out the `setup` module and see if I can come up with something. @cxw620 if you would rather I don't do this lmk and I'm happy to let you have control here :)

**Review by swlynch99 — COMMENTED** [review]
I'm trying to get back up to speed on what's changed here and ran into some stuff I had questions about. This is not an in-depth review. Instead it is mainly just things I am confused about.

File: ktls-util/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,18 @@
+[package]
+name = "ktls-util"
```
[the remark this answers is not in the document]
Comment (cxw620) [reply ktls-util/Cargo.toml:2]: I think we shall contribute these codes to tokio-rustls instead of a new crate, and then implement the fallback mechanism, so than we can benifit from kTLS transparently.

This crate is here for examples and tests only (oh, maybe I should move the codes to `ktls-test`)

File: ktls/Cargo.toml [hunk 2]
```diff
@@ -1,44 +1,50 @@
...
-aws-lc-rs = ["aws_lc_rs"]                                  # Alias because Cargo features commonly use `-`
-ring = ["rustls/ring", "tokio-rustls/ring"]
-tls12 = ["rustls/tls12", "tokio-rustls/tls12"]
+default = ["tls12", "async-io-tokio"]
+
+# Enable rustls TLS 1.2 support
+tls12 = ["rustls/tls12"]
+
+# Expose some low-level APIs
+raw-api = []
```
[the remark this answers is not in the document]
Comment (cxw620) [reply ktls/Cargo.toml:38]: These APIs are not needed in most cases, but using them incorrectly could cause problems, so I gate them behind the feature `raw-api` to *force* users to read the documentation.

File: ktls-util/Cargo.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (swlynch99) [reply ktls-util/Cargo.toml:2]: Uhh, you should probably open an issue on the `tokio-rustls` repository to check that they are actually interested in that. Either way if you're not intending to submit it as part of this PR then it should be moved out to a separate branch. What's currently in the `ktls-util` crate is something that can very clearly be submitted in a separate PR.

**Comment by cxw620** [comment]
> Sorry for disappearing on the last PR and thanks @cxw620 for picking this up! I'm glad this didn't get stalled completely due to me getting burned out over the summer and I like the new work you've done to finish it :)
> 
> Here's some suggestions for things you can break into separate PRs that should help the maintainers review things more easily:
> 
>     * Create a separate PR with mostly just the crate reorganization changes (e.g. creating the `setup`) module. Possibly multiple if there are multiple independent reorganizations.
> 
>     * The rewrite of the code to probe for supported protocols can be moved to its own PR. The original `KtlsStream` struct doesn't interact with this in any way so it should be possible to do this without affecting much else.
> 
>     * I'm not sure that there's a good way to break `KtlsStream` apart so this might be one big PR, but it may be possible to put it in as a skeleton and build it up using multiple PRs.
>
> I'm going to take a crack at splitting out the `setup` module and see if I can come up with something. @cxw620 if you would rather I don't do this lmk and I'm happy to let you have control here :)
> 
> Edit: I have a branch up here that just includes the `setup` and `error` modules that could be submitted independently. https://github.com/swlynch99/tokio-ktls/tree/extract-setup

Thanks for your advices! I will continue to work on this and cherry-pick your commits.

This PR may be broke into these pieces:

- Refactors of mod `setup`
- Refactors of mod `utils::probe`
- Implements `KtlsStream`. I think we can hardly break this into smaller PRs.
  - Implements Tokio's `AsyncRead` / `AsyncWrite` support for `KtlsStream`.

File: ktls/Cargo.toml (hunk 2 above)
[the remark this answers is not in the document]
Comment (cxw620) [reply ktls/Cargo.toml:38]: If we implement Read / Write / AsyncRead / AsyncWrite for KtlsStream, KtlsStream can serve as a replacement of original TcpStream (or something else) in most cases. Exposing these *raw* APIs are just for something like `splice(2)` (this is also why I'm here trying to continue on your work).

Ah, `setup_tls_*` doesn't need to be exposed at all, and marking `KtlsStream::as_raw` as unsafe is enough. I'll remove this feature flag.

File: ktls/src/lib.rs [hunk 3]
```diff
@@ -1,451 +1,33 @@
...
-                KtlsCipherType::Chacha20Poly1305 => cipher_suite::TLS13_CHACHA20_POLY1305_SHA256,
-            },
-        }
-    }
-}
+pub mod log;
+mod protocol;
+pub mod setup;
+pub mod stream;
+pub mod utils;
```
Comment (swlynch99) [inline ktls/src/lib.rs:30]: Instead of this being `pub` we should just export `CompatibleCipherSuites` in the crate root.

File: ktls/src/stream.rs [hunk 4]
```diff
@@ -0,0 +1,251 @@
+//! See [`KtlsStream`].
+
+#![allow(clippy::module_name_repetitions)]
+
+pub mod context;
+pub mod error;
+pub mod impl_std;
+#[cfg(feature = "async-io-tokio")]
+pub mod impl_tokio;
```
Comment (swlynch99) [inline ktls/src/stream.rs:9]: Neither of these modules actually contain any public types.

File: rustfmt.toml [hunk 5]
```diff
@@ -0,0 +1,21 @@
+edition = "2021"
```
Comment (swlynch99) [inline rustfmt.toml:1]: This change should _definitely_ go in its own PR.

File: ktls/src/stream/error.rs [hunk 6]
```diff
@@ -0,0 +1,88 @@
+//! Error type of `KtlsStream` and related operations.
+
+use std::{fmt, io};
+
+use rustls::{AlertDescription, InvalidMessage, PeerMisbehaved};
+
+#[non_exhaustive]
+#[derive(Debug)]
+/// The error type for `KtlsStream` and related operations.
+pub enum KtlsStreamError {
```
Comment (swlynch99) [inline ktls/src/stream/error.rs:10]: This should really document that it is emitted within an `io::Error` from operations on `KtlsStream`.

**Review by swlynch99 — COMMENTED** [review]
I still haven't been able to fully review this, there's too much code to go through at once. However, here's a more in depth review.

**Comment by swlynch99** [comment]
I'm gonna suggest we try and work through some my comments here before trying to split things off as separate PRs. The end goal should be to get everything in this PR merged in separate piecewise PRs that can be effectively reviewed by the existing maintainers to the point where there's nothing left in this PR. I don't think it is possible to effectively review this PR at once given its size.

File: ktls/src/stream.rs (hunk 4 above)
Comment (cxw620) [reply ktls/src/stream.rs:9]: I moved them to separate modules just to make `stream.rs` looks cleaner 🤔

Comment (swlynch99) [reply ktls/src/stream.rs:9]: The issue isn't that they are in separate modules (that's fine). It's that they are public but have no public types. They should just be private.

**Review by swlynch99 — COMMENTED** [review]
A few more comments before I actually go to bed for real this time :P

**Comment by marcus-sa** [comment]
What's the progress on this?
